pkg/udf/rpc/grpc_reduce.go

```diff
@@ -75,111 +74,104 @@ func (u *GRPCBasedReduce) WaitUntilReady(ctx context.Context) error {
 
 // ApplyReduce accepts a channel of isbMessages and returns the aggregated result
 func (u *GRPCBasedReduce) ApplyReduce(ctx context.Context, partitionID *partition.ID, messageStream <-chan *isb.ReadMessage) ([]*isb.WriteMessage, error) {
-	datumCh := make(chan *reducepb.ReduceRequest)
-	var wg sync.WaitGroup
-	var result *reducepb.ReduceResponse
-	var err error
+	var (
+		result     *reducepb.ReduceResponse
+		err        error
+		errCh      = make(chan error, 1)
+		responseCh = make(chan *reducepb.ReduceResponse, 1)
+		datumCh    = make(chan *reducepb.ReduceRequest)
+	)
 
 	// pass key and window information inside the context
 	mdMap := map[string]string{
 		shared.WinStartTime: strconv.FormatInt(partitionID.Start.UnixMilli(), 10),
 		shared.WinEndTime:   strconv.FormatInt(partitionID.End.UnixMilli(), 10),
 	}
 
-	ctx = metadata.NewOutgoingContext(ctx, metadata.New(mdMap))
+	cctx, cancel := context.WithCancel(ctx)
```

Comment:
why do you need a new context?

Reply:
to exit from readloop go routine when we return from ApplyReduce.

Reply:
It should be okay to just do `ctx, cancel := context.WithCancel(ctx)`, do not need to give a new `cctx`.

This is a common patten to make sure everyone in this function watches this `ctx` will properly exit if the function returns, no matter if it's expectedly or unexpectedly.